sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+`"sycl-"` in order to avoid collision with non-SYCL IR attributes, but this is
+not a strict requirement.
+
+The property value can be a literal or `constexpr` expression of the following
+types:
+
+* `const char *`.
+* An integer type.
+* A floating point type.
+* A boolean type.
+* A character type.
+* An enumeration type.
+
+All properties require a value when represented in the C++ attribute.  If the
+SYCL property has no value the header passes the empty string (`""`).
+
+The front-end converts each value to a string before representing it in LLVM
+IR.  Integer and floating point values are converted with the same format as
+`std::to_string()` would produce.  Boolean values are converted to either
+`"true"` or `"false"`.  Enumeration values are first converted to an integer
+and then converted to a string with the same format as `std::to_string()`.
+
+**TODO**: Should we allow property values that are type names?  If so, I
+suppose they would be converted to a string representation of the mangled name?
+
+**TODO**: Should we allow property values of other (non-fundamental) types?  If
+we allow this, we need to teach the front-end how to convert each type to a
+string, which means the front-end needs to be changed each time we add a
+property with a new non-fundamental type.  This seems undesirable.  However, if
+we do not allow non-fundamental types, how do we represent properties like
+`work_group_size`, whose value is a 3-tuple of integers?  Maybe we could just
+allow `std::tuple`, where the type of each element is one of the fundamental
+types listed above.
+
+
+## Representing properties in SPIR-V
+
+There is no mechanical process which converts all LLVM IR attributes to
+SPIR-V.  This is because we do not need all properties to be expressed in
+SPIR-V and because there is no consistent way to represent properties in
+SPIR-V.  Therefore, the `sycl-post-link` tool decides on a case-by-case basis
+which properties are translated into SPIR-V and which representation to use.
+
+We use the [SPIR-V LLVM Translator][11] to translate from LLVM IR to SPIR-V,
+and that tool defines [idiomatic LLVM IR][12] representations that correspond
+to various SPIR-V instructions.  Therefore, the `sycl-post-link` tool can
+translate a property into a specific SPIR-V instruction by generating the
+corresponding idiomatic LLVM IR.  The following sections describe some common
+cases.
+
+[11]: <https://github.com/KhronosGroup/SPIRV-LLVM-Translator>
+[12]: <https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/docs/SPIRVRepresentationInLLVM.rst>
+
+### Property on a kernel function
+
+When a property on a kernel function needs to be represented in SPIR-V, we
+generally translate the property into a SPIR-V **OpExecutionMode** instruction.
+The SPIR-V LLVM Translator has an existing way to generate this instruction
+when the LLVM IR contains the named metadata `!spirv.ExecutionMode` as
+illustrated below:
```

Comment:
It seems like we need some kind of a mapping to be recorded in some component of the toolchain to convert properties into numbers which correspond to execution modes in SPIR-V, right?

Where do you think such mapping should be placed? Will it be placed within sycl-post-link, DPC++ FE or headers?

Reply:
Yes, I expect this logic to be in `sycl-post-link`.  The introduction to this section says:

> Therefore, the `sycl-post-link` tool decides on a case-by-case basis which properties are translated into SPIR-V and which representation to use.

I could expand that if you think it needs more clarification.

Reply:
> Yes, I expect this logic to be in `sycl-post-link`

That means that every time new property is added, we still have to update the compiler (some parts of it) to properly support it, right? And even if we move that mapping into RT, then we will anyway have to update our SPIR-V translator if we introduce new property which can't be represented with existing SPIR-V capabilities

Reply:
Even the headers are part of the compiler release, though, so we will certainly need to change *some* part of the compiler whenever we add a new property.  I thought it should be a goal to avoid the need to change the front-end when we add new properties since that seems to be a particular pain point.  It also seemed easy to design this in a way that does not rely a change to the "SPIR-V LLVM Translator" each time we need add a new property that needs representation in SPIR-V.

So I these are the parts of the compiler that will need to change when we add a new property:

* The header file that defines the property and the header that defines the class / function that uses the property.
* The IR compiler phase that uses the property, if any.  (For example, this might be the `sycl-post-link` tool.)
* If the property needs to be represented in SPIR-V:
  - The `sycl-post-link` tool, to translate the property into idiomatic IR for the "SPIR-V LLVM Translator".
  - The SPIR-V consumer (e.g. JIT compiler), which uses the property.